### ISSUE TITLE
Duplicate logs when REST API returns HTTP Status Code 401 or 403.

### ISSUE THREAD

**Issue by srknkbs** [zalando/logbook#656]
Duplicate logs when REST API returns HTTP Status Code 401 or 403.

## Description
We have a REST API that returns HTTP Status Code 401 or 403 when the user does not have permission or authentication issues detected inside the API.

When we look at the Logbook code, we see that there is a "SecurityStrategy" used for filtering unauthorized request before it comes to REST API. So, Logbook works as expected if we try to call REST API with invalid token which means Secure filter works as designed.

On the other hand, if I call the REST API with a user who does not have correct permissions, our Spring Security part passes and REST API returns HTTP Status Code 401 when it detects unauthorized access. In this case, Logbook generates duplicate records since first Default Logbook Filter is being executed and then it comes to Secure Logbook Filter and generates same logs again.

https://github.com/zalando/logbook/blob/668344ee75937820f1f33640fd3df3d66d57065a/logbook-core/src/main/java/org/zalando/logbook/SecurityStrategy.java#L23-L35

## Expected Behavior
One log for request and one log for response

## Actual Behavior
Two logs for request and two logs for response

## Possible Fix
There is a workaround. We can skip Secure Logbook filter if we set flag "logbook.secure-filter.enabled" as false. But in this case, we have logs with bodies (unauthorized requests).

## Steps to Reproduce
1. Create a Spring Boot Application with a REST API which validates user roles and returns 401 or 403 when permission denied.
2. Add Logbook dependency to your Spring Boot Application
3. Call REST API with unauthorized user
4. Observe in the Console Log that Logbook generates duplicate logs.

## Context
We have duplicate records in our trace database when REST API returns 401 or 403.

**Comment by whiskeysierra** [comment]
As you correctly identified the `SecureLogbookFilter` and the normal `LogbookFilter` work together based on the assumption that any security-related logic is done in a filter that is running after `SecureLogbookFilter` but before `LogbookFilter`.

Any chance you can move your security logic into a filter?

**Comment by srknkbs** [comment]
Yes, I can try it, but what i could not understand that when it comes to Logbookfilter, then it should not come to SecureLogbookFilter. 
But in my case, since my REST API returns 401 it generates log this way:

* generate Log for request (DefaultStrategy)
* execute REST API logic (returns 401)
* generate Log for response (DefaultStrategy)
* generate Log for request (SecurityStrategy)
* generate Log for request (SecurityStrategy)

**Comment by whiskeysierra** [comment]
1. `SecureLogbookFilter` delays logging of request
2. `LogbookFilter` logs request (with body, since assumes it's authorized by now)
3. API returns 401
4. `LogbookFilter` logs response
5. `SecureLogbookFilter` now logs request late since the status code is 401 (without body, to not flood the logs)
6. `SecureLogbookFilter` now logs response since it assumes no one else did log it

**Comment by whiskeysierra** [comment]
So the behavior of the two filters matches exactly my expectation in this case. Their are fundamentally incompatible with your setup, unfortunately. Either you change your application or we need to find another way to achieve what you need using potentially new constructs (maybe a custom filter that combines both behaviors?) or two filters that communicate with each other using attributes (I tried a lot to avoid that until now).

**Comment by srknkbs** [comment]
I understand your points and the design constraints for `Logbookfilters`. However, there might always 
be a case that two `Logbookfilter` overlap and generate duplicate logs.

**Comment by whiskeysierra** [comment]
> However, there might always
> be a case that two `Logbookfilter` overlap and generate duplicate logs.

They won't as long as the only one producing 401/403 responses is a filter in between them. For all other scenarios this setup is not intended and won't work.

**Comment by whiskeysierra** [comment]
The introduction of the `Strategy` interface was primarily driven to ease customization efforts for cases like this one. 

The following strategy would:

- log unauthorized (response status code 401/403) requests without body
- have same behavior whether it's Spring Security (filter) or your own application/servlet/controller responding with 401/403
- always delay request logging until response is available (necessary since it's impossible to determine authorized vs unauthorized before)
- require to be running in a **single filter** but **before** Spring's SecurityFilter

```java
public final class CustomStrategy implements Strategy {

    @Override
    public void write(final Precorrelation precorrelation, final HttpRequest request, final Sink sink) {
        // defer request logging until response is available
    }

    @Override
    public void write(final Correlation correlation, final HttpRequest request, final HttpResponse response,
            final Sink sink) throws IOException {

        final int status = response.getStatus();
        if (status == 401 || status == 403) {
            sink.writeBoth(correlation, request.withoutBody(), response);
        } else {
            sink.writeBoth(correlation, request, response);
        }
    }

}
```

**Comment by srknkbs** [comment]
> 1. `SecureLogbookFilter` delays logging of request
> 2. `LogbookFilter` logs request (with body, since assumes it's authorized by now)
> 3. API returns 401
> 4. `LogbookFilter` logs response
> 5. `SecureLogbookFilter` now logs request late since the status code is 401 (without body, to not flood the logs)
> 6. `SecureLogbookFilter` now logs response since it assumes no one else did log it

Is there any way to filter responses between step 4 and 5 or stop the process?

**Comment by Sam-Kruglov** [comment]
Hi @whiskeysierra, I have a similar case but it just writes response twice every time (request is only written once). In the debugger, I saw the first log was done by HttpFilter and second by HttpFilter > SecureLogbookFilter. This happens for 200 responses even

**Comment by whiskeysierra** [comment]
> Is there any way to filter responses between step 4 and 5 or skip step 5 and 6 somehow?
> 
> I asked this becaues we have a requirement that we can not delay request logs for authorized operations.

@srknkbs You would need to write your own two filters that communicate with each other using request attributes.

> This happens for 200 responses even.

@Sam-Kruglov That is rather unexpected. Can you drop the log statements here (obfuscated, if needed)?

**Comment by Sam-Kruglov** [comment]
@whiskeysierra I just also witnessed that after introducing Logbook, my custom authorization `javax.servlet.Filter` is now called twice: before processing and before returning, where second time is unexpected (since the request is already authorized and even processed). I will try to dig some more but let me know if you need anything. 

Not sure what logs do you want to see? It just prints "Incoming Request" and then twice "Outgoing Response", maybe some logbook debug level or trace level things?

**Comment by whiskeysierra** [comment]
Can you prepare a reproducible test case or demo project that suffers from this?

**Comment by Sam-Kruglov** [comment]
sorry, just removed logbook entirely, it still does it, not sure what is going on

**Comment by Sam-Kruglov** [comment]
@whiskeysierra found it, it's because I return `Mono` from my Spring MVC controller, so it does async dispatch. I fixed my authentication filter by using advice from here https://stackoverflow.com/a/48322641/6166627

Logbook still prints my response twice. But if I set `logbook.secure-filter.enabled=false` then only once, as expected. I suppose that `SecurityLogbookFilter` should somehow extend `OncePerRequestFilter`

**Comment by whiskeysierra** [comment]
There is an open pull request to fix async dispatch.

**Comment by whiskeysierra** [comment]
@srknkbs 

> But in my case, since my REST API returns 401 it [...]

and

> we have a requirement that we can not delay request logs for authorized operations

Those two requirements are impossible to satisfy at the same time. If your controller is doing authorization then you have to delay logging of the request because only the response will tell you whether the request was authorized or not.

**Comment by whiskeysierra** [comment]
@Sam-Kruglov Fix for broken async dispatch has been released as 2.0.0-RC.7

**Comment by whiskeysierra** [comment]
@srknkbs I don't see how your requirements could work. At least it's not supported with the current set of filters. Feel free to reopen this if you believe there is a suitable extension/feature for logbook.
